util/hujson/types.go

```diff
@@ -0,0 +1,398 @@
+
+// Float construct a JSON literal for a floating-point number.
+// The values NaN, +Inf, and -Inf will be represented as a JSON string
+// with the values "NaN", "Infinity", and "-Infinity".
+func Float(v float64) Literal {
+	switch {
+	case math.IsNaN(v):
+		return Literal(`"NaN"`)
+	case math.IsInf(v, +1):
+		return Literal(`"Infinity"`)
+	case math.IsInf(v, -1):
+		return Literal(`"-Infinity"`)
+	default:
+		b, _ := json.Marshal(v)
+		return Literal(b)
+	}
+}
+
+// Clone makes a new copy of the literal.
+func (b Literal) Clone() value {
+	return Literal(copyBytes(b))
+}
+
+// Kind represents each possible JSON literal kind with a single byte,
+// which is conveniently the first byte of that kind's grammar
+// with the restriction that numbers always be represented with '0'.
+func (b Literal) Kind() byte {
+	if len(b) == 0 {
+		return 0
+	}
+	switch k := b[0]; k {
+	case 'n', 'f', 't', '"':
+		return k
+	case '-', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9':
+		return '0'
+	default:
+		return 0
+	}
+}
+
+// IsValid reports whether b is a valid JSON null, boolean, string, or number.
+// The lieral must not have surrounding whitespace.
+func (b Literal) IsValid() bool {
+	// NOTE: The v1 json package is non-compliant with RFC 8259, section 8.1
+	// in that it does not enforce the use of valid UTF-8.
+	return json.Valid(b) && len(b) == len(bytes.TrimSpace(b))
+}
+
+// Bool returns the value for a JSON boolean.
+// It returns false if the literal is not a JSON boolean.
+func (b Literal) Bool() bool {
+	return string(b) == "true"
+}
+
+// String returns the unescaped string value for a JSON string.
+// For other JSON kinds, this returns the raw JSON represention.
+func (b Literal) String() (s string) {
+	if b.Kind() == '"' && json.Unmarshal(b, &s) == nil {
+		return s
+	}
+	return string(b)
+}
+
+// Int returns the signed integer value for a JSON number.
+// It returns 0 if the literal is not a signed integer.
+func (b Literal) Int() (n int64) {
+	if b.Kind() == '0' && json.Unmarshal(b, &n) == nil {
+		return n
+	}
+	return 0
+}
+
+// Uin returns the unsigned integer value for a JSON number.
+// It returns 0 if the literal is not an unsigned integer.
+func (b Literal) Uint() (n uint64) {
+	if b.Kind() == '0' && json.Unmarshal(b, &n) == nil {
+		return n
+	}
+	return 0
+}
+
+// Float returns the floating-point value for a JSON number.
+// It returns a NaN, +Inf, or -Inf value for any JSON string with the values
+// "NaN", "Infinity", or "-Infinity".
+// It returns 0 for all other cases.
+func (b Literal) Float() (n float64) {
+	if b.Kind() == '0' && json.Unmarshal(b, &n) == nil {
+		return n
+	}
+	if b.Kind() == '"' {
+		switch b.String() {
+		case "NaN":
+			return math.NaN()
+		case "Infinity":
+			return math.Inf(+1)
+		case "-Infinity":
+			return math.Inf(-1)
+		}
+	}
+	return 0
+}
+
+// Object is an exact syntactic representation of a JSON object.
+type Object struct {
+	// Members are the members of a JSON object.
+	Members [][2]Value
```

Comment:
can you make `[2]Value` a named type so docs can be attached to it? What are the values at index 0 and 1? etc.

Reply:
Done.